File: python/aibrix/aibrix/planner/api/v1/planner.py
# Copyright 2025 The Aibrix Team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# 	http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Dict, List, Optional

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel, Field

from aibrix.logger import init_logger
from aibrix.planner.errors import PlannerError
from aibrix.planner.models import (
    ModelGPUProfile,
    PlannerJob,
    SchedulingDecision,
)
from aibrix.planner.scheduler import Scheduler
from aibrix.planner.snapshot import PlannerSnapshot

logger = init_logger(__name__)

router = APIRouter()


class ResourceDeliveryEvent(BaseModel):
    reservation_id: str
    region_id: Optional[str] = None
    cluster_id: str
    gpu_type: str
    requested_cards: int
    delivered_cards: int
    start_hour: int
    duration_hours: int
    status: str


class PlannerStatusResponse(BaseModel):
    mode: str
    planner_loop_enabled: bool
    rm_base_url: str


class PlanTriggerContext(BaseModel):
    trigger: str = Field(default="manual", description="manual|periodic|event")
    event_type: Optional[str] = None
    request_id: Optional[str] = None


class PlanRequest(BaseModel):
    jobs: List[PlannerJob] = Field(default_factory=list)
    profiles: List[ModelGPUProfile] = Field(default_factory=list)
    context: Optional[PlanTriggerContext] = None


class PlanResponse(BaseModel):
    decisions: List[SchedulingDecision] = Field(default_factory=list)
    total_jobs: int
    trigger: str


def _get_scheduler(request: Request) -> Scheduler:
    if not hasattr(request.app.state, "planner_scheduler"):
        raise HTTPException(status_code=503, detail="Planner not initialized")
    return request.app.state.planner_scheduler


@router.post("/events/resource-delivery")
async def resource_delivery(request: Request, _event: ResourceDeliveryEvent) -> Dict:
    _get_scheduler(request)
    raise HTTPException(
        status_code=501,
        detail=(
            "Resource-delivery webhook ingest is not implemented in MVP. "
            "The planner observes RM reservation state by polling on each "
            "/v1/planner/plan call; a push-based receiver is on the deferred "
            "roadmap (see MVP_DESIGN.md)."
        ),
    )


@router.post("/plan")
async def plan(request: Request, payload: PlanRequest) -> Dict:
    scheduler = _get_scheduler(request)
    snapshot = PlannerSnapshot(
        jobs=payload.jobs,
        profiles=payload.profiles,
    )
    try:
        decisions = await scheduler.plan(snapshot)
    except PlannerError as e:
        # Translate typed planner errors into accurate HTTP status codes so
        # callers can distinguish "nothing to do" (200 + empty list) from a
        # degraded collaborator (e.g. RM unavailable -> 503).
        logger.warning(
            "Plan cycle aborted with PlannerError",
            error=str(e),
            type=type(e).__name__,
        )
        raise HTTPException(status_code=e.http_status, detail=str(e)) from e
    trigger = payload.context.trigger if payload.context else "manual"
    return PlanResponse(
        decisions=decisions,
        total_jobs=len(payload.jobs),
        trigger=trigger,
    ).model_dump()


@router.post("/schedule")
async def trigger_schedule(request: Request) -> Dict:
    _get_scheduler(request)
    raise HTTPException(
        status_code=501,
        detail=(
            "Planner no longer self-schedules. Let MDS call POST /v1/planner/plan "
            "periodically or on events with the current job snapshot."
        ),
    )


@router.get("/status")
async def planner_status(request: Request) -> Dict:
    scheduler = _get_scheduler(request)
    return PlannerStatusResponse(
        mode="stateless-request-driven",
        planner_loop_enabled=False,
        rm_base_url=scheduler.rm_client.base_url,
    ).model_dump()

File: apps/console/api/handler/planner.go
/*
Copyright 2025 The Aibrix Team.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package handler

import (
	"context"
	"encoding/json"
	"errors"
	"io"
	"net/http"

	"github.com/grpc-ecosystem/grpc-gateway/v2/runtime"
	"k8s.io/klog/v2"

	"github.com/vllm-project/aibrix/pkg/planner"
)

// PlannerScheduler is the subset of *planner.Scheduler the handler needs.
// Defined as an interface so handler tests can substitute a fake without
// constructing a real scheduler with stub collaborators.
type PlannerScheduler interface {
	Schedule(ctx context.Context, job planner.PlannerJob) (*planner.SchedulingDecision, error)
}

// PlannerHandler exposes the planner over HTTP. It is a thin adapter: it
// parses JSON, calls Scheduler.Schedule, and translates planner errors to
// HTTP status codes.
//
// This handler uses the same mux.HandlePath registration style as
// FileHandler and PlaygroundHandler. When the gRPC migration lands it will
// be replaced by a generated PlannerService server; the planner package
// itself will not change.
type PlannerHandler struct {
	scheduler PlannerScheduler
}

// NewPlannerHandler constructs a PlannerHandler around the given scheduler.
func NewPlannerHandler(scheduler PlannerScheduler) *PlannerHandler {
	return &PlannerHandler{scheduler: scheduler}
}

// RegisterRoutes attaches planner routes to the grpc-gateway mux.
func (h *PlannerHandler) RegisterRoutes(mux *runtime.ServeMux) {
	if err := mux.HandlePath("POST", "/api/v1/planner/jobs", h.handleSubmit); err != nil {
		klog.Fatalf("Failed to register planner submit route: %v", err)
	}
	if err := mux.HandlePath("GET", "/api/v1/planner/status", h.handleStatus); err != nil {
		klog.Fatalf("Failed to register planner status route: %v", err)
	}
}

// submitJobRequest is the JSON shape accepted by POST /api/v1/planner/jobs.
type submitJobRequest struct {
	JobID         string          `json:"job_id"`
	ModelID       string          `json:"model_id,omitempty"`
	GPUType       string          `json:"gpu_type"`
	GPUCount      int             `json:"gpu_count"`
	StartHour     int64           `json:"start_hour"`
	DurationHours int             `json:"duration_hours"`
	RegionID      string          `json:"region_id,omitempty"`
	BatchPayload  batchPayloadDTO `json:"batch_payload"`
}

type batchPayloadDTO struct {
	InputFileID      string            `json:"input_file_id"`
	Endpoint         string            `json:"endpoint"`
	CompletionWindow string            `json:"completion_window,omitempty"`
	Metadata         map[string]string `json:"metadata,omitempty"`
	// ModelDeploymentTemplate selection. The metadata service requires a
	// template to render the K8s Job; submissions without one are rejected
	// at render time. Empty Version means "latest active version of Name".
	ModelTemplateName    string `json:"model_template_name,omitempty"`
	ModelTemplateVersion string `json:"model_template_version,omitempty"`
}

// submitJobResponse is the JSON shape returned by POST /api/v1/planner/jobs.
type submitJobResponse struct {
	JobID         string `json:"job_id"`
	Status        string `json:"status"`
	ReservationID string `json:"reservation_id,omitempty"`
	ClusterID     string `json:"cluster_id,omitempty"`
	BatchID       string `json:"batch_id,omitempty"`
}

type plannerStatusResponse struct {
	Mode    string `json:"mode"`
	Adapter string `json:"adapter"`
}

type plannerErrorResponse struct {
	Error string `json:"error"`
}

func (h *PlannerHandler) handleSubmit(w http.ResponseWriter, r *http.Request, _ map[string]string) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		writePlannerError(w, http.StatusBadRequest, "failed to read request body")
		return
	}
	defer func() { _ = r.Body.Close() }()

	var req submitJobRequest
	if err := json.Unmarshal(body, &req); err != nil {
		writePlannerError(w, http.StatusBadRequest, "invalid json: "+err.Error())
		return
	}

	job := planner.PlannerJob{
		JobID:         req.JobID,
		ModelID:       req.ModelID,
		GPUType:       req.GPUType,
		GPUCount:      req.GPUCount,
		StartHour:     req.StartHour,
		DurationHours: req.DurationHours,
		RegionID:      req.RegionID,
		BatchPayload: planner.BatchPayload{
			InputFileID:          req.BatchPayload.InputFileID,
			Endpoint:             req.BatchPayload.Endpoint,
			CompletionWindow:     req.BatchPayload.CompletionWindow,
			Metadata:             req.BatchPayload.Metadata,
			ModelTemplateName:    req.BatchPayload.ModelTemplateName,
			ModelTemplateVersion: req.BatchPayload.ModelTemplateVersion,
		},
	}

	decision, err := h.scheduler.Schedule(r.Context(), job)
	if err != nil {
		writePlannerError(w, plannerErrorStatus(err), err.Error())
		return
	}

	writePlannerJSON(w, http.StatusOK, submitJobResponse{
		JobID:         decision.JobID,
		Status:        string(decision.Status),
		ReservationID: decision.ReservationID,
		ClusterID:     decision.ClusterID,
		BatchID:       decision.BatchID,
	})
}

func (h *PlannerHandler) handleStatus(w http.ResponseWriter, _ *http.Request, _ map[string]string) {
	writePlannerJSON(w, http.StatusOK, plannerStatusResponse{
		Mode:    "in-process",
		Adapter: "stub",
	})
}

// plannerErrorStatus maps a typed planner error to its HTTP status code.
// Falls back to 500 for unexpected errors so misclassification surfaces in logs.
func plannerErrorStatus(err error) int {
	switch {
	case errors.Is(err, planner.ErrInvalidJob):
		return http.StatusBadRequest
	case errors.Is(err, planner.ErrCapacityUnavailable):
		return http.StatusConflict
	case errors.Is(err, planner.ErrRMUnavailable):
		return http.StatusServiceUnavailable
	case errors.Is(err, planner.ErrMDSSubmitFailed):
		return http.StatusBadGateway
	default:
		return http.StatusInternalServerError
	}
}

func writePlannerJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	if err := json.NewEncoder(w).Encode(body); err != nil {
		klog.Errorf("planner: failed to encode response: %v", err)
	}
}

func writePlannerError(w http.ResponseWriter, status int, msg string) {
	writePlannerJSON(w, status, plannerErrorResponse{Error: msg})
}
